client-v2/src/shared/util/shared.ts

```diff
@@ -26,38 +26,78 @@ const createGroup = (
 
 const getUUID = <T extends { uuid: string }>({ uuid }: T) => uuid;
 
-// this is horrible but is the only way to do it!
-const groupByIndex = (groups: Group[], index: number): Group | undefined =>
-  groups.find(g => parseInt(g.id || '0', 10) === index);
+const getGroupIndex = (id: string | null): number => parseInt(id || '0', 10);
 
 const getAllArticleFragments = (groups: Group[]) =>
   groups.reduce(
     (acc, { articleFragments }) => [...acc, ...articleFragments],
     [] as string[]
   );
 
-const addEmptyGroupsFromCollectionConfigForStage = (
+const configGroupIndexExistsInGroups = (
+  groupsToSearch: Group[],
+  index: number
+): boolean =>
+  groupsToSearch.some(group => {
+    if (group.id) {
+      return parseInt(group.id, 10) === index;
+    }
+    return index === 0;
+  });
+
+const addGroupsForStage = (
   groupIds: string[],
   entities: { [id: string]: Group },
   collectionConfig: CollectionConfig
 ) => {
   const groups = groupIds.map(id => entities[id]);
-  const addedGroups = collectionConfig.groups
-    ? collectionConfig.groups.map((name, i) => {
-        const existingGroup = groupByIndex(groups, i);
+  let groupsWithNames = groups.map(group => {
+    let name: string | null = null;
+    const groupNumberAsInt = getGroupIndex(group.id);
+    if (
+      collectionConfig.groups &&
+      groupNumberAsInt < collectionConfig.groups.length
+    ) {
+      name = collectionConfig.groups[groupNumberAsInt];
+    }
+    return { ...group, name };
+  });
+
+  // We may have empty groups in the config which would not show up in the normalised
+  // groups result. We need to add these into the groups array.
+  if (collectionConfig.groups) {
+    collectionConfig.groups.forEach((group, configGroupIndex) => {
+      if (!configGroupIndexExistsInGroups(groupsWithNames, configGroupIndex)) {
+        groupsWithNames.push(createGroup(`${configGroupIndex}`, group));
+      }
+    });
+  }
+
+  // If we have no article fragments and no groups in a collection we still need to create
+  // and empty group for articles.
+  if (groupsWithNames.length === 0) {
+    groupsWithNames = [createGroup(null, null, getAllArticleFragments(groups))];
+  }
+
+  // Finally we need to sort the groups according to their ids.
+  groupsWithNames.sort((g1, g2) => {
```

Comment:
Only a suggestion, but you could remove ten+ lines in a long-ish function by using lodash's sortBy, which accepts a predicate